r-pkg/R/helperfuns.R

```diff
@@ -21,28 +32,135 @@
     )
 }
 
+# [title] List out HTTP codes that should be treated as retryable
+# [name] .should_retry
+# [description] Here because {curl} doesn't ship a retry mechanism, so this library
+#               implements its own.
+.should_retry <- function(response) {
+    retryable_error_codes <- c(
+        # 408 - timeout
+        408L
+        # 422 - unprocessable entity
+        , 422L
+        # 425 - too early
+        , 425L
+        # 429 - too many requests
+        , 429L
+        # 500 - internal server error
+        , 500L
+        # 502 - bad gateway
+        , 502L
+        # 503 - service unavailable
+        , 503L
+        # 504 - gateway timeout
+        , 504L
+    )
+    return(response$status_code %in% retryable_error_codes)
+}
+
+# [title] Retry an HTTP requests a couple times (if necessary)
+# [name] .retry
+# [description] Implements exponential backoff with jitter, around failed requests.
+#               See .should_retry() for details on which status codes are considered retryable.
+#               This is here because {curl} does not have a built-in retry API.
+#' @importFrom curl curl_fetch_memory
+#' @importFrom stats runif
+.retry <- function(handle, url) {
+
+    max_retries <- 3L
+    attempt_count <- 1L
+    while (attempt_count <= max_retries) {
+
+        # if this isn't the 1st attempt, apply backoff
+        if (attempt_count > 1L) {
+            # exponential backoff with jitter
+            #
+            #   1.45s + {jitter}
+            #   2.10s + {jitter}
+            #   3.05s + {jitter}
+            #   etc., etc.
+            #
+            # ref: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+            sleep_seconds <- 1.45 ** (attempt_count - 1L) + stats::runif(n = 1L, min = 0.1, max = 0.5)
```

Comment:
1.45s and the jitter being 0.1-0.5 are arbitrary. @austin3dickey if you have a different suggestion for these I'd take them, I really don't feel that strongly about it.

Reply:
nah this seems in line with what's normally done!

Reply:
ok great, thank you